### ISSUE TITLE
Use descriptive keyword argument instead of `t` when boolean flag is expected

### ISSUE THREAD

**Issue by Fuco1** [bbatsov/emacs-lisp-style-guide#30]
For example, functions which expects 5 arguments, all of which are `t` can get very confusing with calls like `(foo t nil t nil t)`. It is not obvious what to do.

Recently I've been naming such arguments with descriptive names and then pass `:argument-name` keyword as a value. For example

``` scheme
(defun update (&optional force)
   "Update the list of values. If FORCE is non-nil, invalidate cache first.")
```

You would then call it `(update)` or `(update :force)` instead of `(update t)`.

I think there is very little confusion possible over keyword-argument-functions and keyword-as-flags usage, especially when you can consult the documentation in about 1 second.

What do you think people?

**Comment by bbatsov** [comment]
Ping :-)

**Comment by Fuco1** [comment]
One advantage of `:foo` over `'foo` is that they are fontified differently.  For about a year now I use a special rule to fontify constants, such as `'foo` so that helps (for me personally).

I don't mind either way, `:arg` or `'arg` is fine, I would still slightly prefer `:arg`.

**Comment by Fuco1** [comment]
@bbatsov Is this project still alive? Can you assign this to me, I'll add a note somewhere about this.  Let's go with the `'force` style, it is used in Emacs itself which is a good point and worth following.

**Comment by bbatsov** [comment]
Yeah, it is. And every month I plan to come back to it and write some many things, but I'm buried with work, life and side projects. I've made you a collaborator, so feel free to make things happen! :-)

**Comment by Fuco1** [comment]
@bbatsov I have a secret project writing an elisp static analyser kind of thing... so I would eventually like to include these rules as a plugin there.  But I'm also not making a lot of progress (https://github.com/Fuco1/Elsa)

**Comment by bbatsov** [comment]
Nice! 

And I know that @gonewest818 has been working on reviving `elint` recently. I'd be happy to help as well, but that's unlikely to happen any time soon.

**Comment by gonewest818** [comment]
@Fuco1:  fwiw, I adopted `elisp-lint` from the original author (who has since moved onto other things) and did a little spring cleaning on it a few months back. You'll see it's fairly limited in what it does -- either doing string matches on the text file or, in some cases, invoking existing emacs "infrastructure" packages like `bytecomp`, `checkdoc`, `check-declare`, or `package` to validate the code.  There's definitely room for more/better tools in this area.  If you see anything that can/ought to be changed in `elisp-lint` please let me know.

**Comment by Fuco1** [comment]
@gonewest818 I have a completely different approach.  What I am doing is reading the forms, parsing the structure and actually running type inference to see what is going on.  You can annotate the forms where types are ambiguous to help the engine to do a better job.  You get a lot of information like the scope, variables, globals, functions and all sorts of other things at each point in the AST so the decisions are much more informed.

So this will be 100% syntactically based which allows for massively broader class of checks.  The linting stuff would only piggyback on the existing infrastructure but that's not the main goal.
